File: 백준/Silver/9012. 괄호/괄호.py
import sys

input = sys.stdin.readline
t = int(input())

def isVPS(ps):
  stack = []
  for p in ps:
    if not stack and p == ')':
      return 'NO'
    if p == '(':
      stack.append(p)
    if p == ')':
      stack.pop()
  if stack:
    return 'NO'
  else:
    return 'YES'

for _ in range(t):
  ps = list(input().strip())
  print(isVPS(ps))

File: 백준/Silver/9012. 괄호/괄호.java
import java.util.Scanner;
import java.util.Stack;

public class Main {
	public static void main(String[] args) {
		Scanner sc = new Scanner(System.in);
		int t = sc.nextInt();
		for (int tc = 0; tc < t; tc++) {
			String ps = sc.next();
			Stack<Character> stack = new Stack<>();
			boolean flag = true;

			for (int i = 0; i < ps.length(); i++) {
				char p = ps.charAt(i);
				if (stack.isEmpty() && p == ')') {
					flag = false;
				} else {
					if (p == '(') {
						stack.add(p);
					} else {
						stack.pop();
					}
				}
			}
			if (flag && stack.isEmpty()) {
				System.out.println("YES");
			} else {
				System.out.println("NO");
			}
		}
		sc.close();
	}
}